### PR TITLE
Add bash script to autotest cpython lib

### PR DESCRIPTION
**Pull request by MegasKomnenos**
https://github.com/RustPython/RustPython/issues/4564

The following PR contains a bash script to automate the initial testing workflow for updating python library components from cpython. 

The workflow includes selecting and updating an individual library component, then running the related test to search for bugs. I realized that this work can be automated and be run for every library components, to screen out ones that fail. I wrote a bash script that does that, and made this PR.

### CODE REVIEW

File: scripts/clib/readme.txt [hunk 1]
```diff
@@ -0,0 +1,11 @@
+To use, write down your local cpython path in clib_path.txt, and run,
+bash clib_test_one.sh
+
+and check the result in clib_out.txt
+
+The script will try to test every component in clib_list.txt, where,
+
+clib_test_one.sh : test each component individually
+clib_test_all.sh : test every components simultaneously
```
Comment (fanninpm) [inline scripts/clib/readme.txt:9]: Grammar nitpick (adjective-noun agreement):
```suggestion
clib_test_all.sh : test all components simultaneously
```
or:
```suggestion
clib_test_all.sh : test every component simultaneously
```

File: scripts/clib/clib_test_all.sh [hunk 2]
```diff
@@ -0,0 +1,68 @@
+#!/bin/bash
```
Comment (youknowone) [inline scripts/clib/clib_test_all.sh:1]: shell is good enough, but if possible, can this script be a python script? Practically, we usually struggle with windows tests because its different from unix.
But running bash script on windows is not easy.

Comment (DimitrisJim) [reply scripts/clib/clib_test_all.sh:1]: plus maintainability. Sure we could get issues fixed but if possible we should limit the # of languages we use (as much as possible).

File: scripts/clib/clib_path.txt
Comment (fanninpm) [inline scripts/clib/clib_path.txt:1]: I'm not sure this file should be checked into the repository, as it could vary from user to user.

File: scripts/clib/readme.txt [hunk 3]
```diff
@@ -0,0 +1,11 @@
+To use, write down your local cpython path in clib_path.txt, and run,
```
Comment (DimitrisJim) [inline scripts/clib/readme.txt:1]: could better yet be an `.md` for auto rendering by github.

File: scripts/clib/readme.txt (hunk 1 above)
Comment (MegasKomnenos) [reply scripts/clib/readme.txt:9]: Hehe, thanks! Grammar has never been my strong suit :D

File: scripts/clib/clib_test_all.sh (hunk 2 above)
Comment (MegasKomnenos) [reply scripts/clib/clib_test_all.sh:1]: I think that could be doable, as python supports file IO system calls, and I think it should also be possible to use os.system() to run cargo for building crates.

One question, would that python script be run in RustPython, or just python?

File: scripts/clib/clib_list.txt [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+warnings
+weakref
+webbrowser
+xdrlib
+zipapp
+zipfile
+zipimport
+
+# Components listed below are excluded from being tested
+signal
```
[the remark this answers is not in the document]
Comment (MegasKomnenos) [reply scripts/clib/clib_list.txt:156]: As in, something like this?

shutil
# signal
threading

File: scripts/clib/clib_path.txt
Comment (MegasKomnenos) [reply scripts/clib/clib_path.txt:1]: That is a fair point, it is indeed something that would vary from user to user. I thought I'd leave it there as a sort of guideline for someone trying to use it, but I guess that could also go into readme.

File: scripts/clib/readme.txt (hunk 3 above)
Comment (MegasKomnenos) [reply scripts/clib/readme.txt:1]: :+1:

File: scripts/clib/clib_list.txt (hunk 4 above)
[the remark this answers is not in the document]
Comment (fanninpm) [reply scripts/clib/clib_list.txt:156]: Yes, exactly.

[the remark this answers is not in the document]
Comment (MegasKomnenos) [reply scripts/clib/clib_list.txt:156]: I can see why that would be advantageous. I implemented it like this mainly for ease of implementation, but that is a better way. Let me work on the fix real quick.

[the remark this answers is not in the document]
Comment (MegasKomnenos) [reply scripts/clib/clib_list.txt]: Done!

File: scripts/clib/clib_path.txt
Comment (youknowone) [reply scripts/clib/clib_path.txt:1]: If you are going to use python, argparse is a good tool to manage command line arguments

Comment (youknowone) [reply scripts/clib/clib_path.txt:1]: I agree about it. Adding a default argument will be a good option.

Comment (youknowone) [reply scripts/clib/clib_path.txt:1]: Or just printing a sample command is another good option. User can learn from error messages quickly.

Comment (MegasKomnenos) [reply scripts/clib/clib_path.txt:1]: I think both can work together, where there's a default argument value and an error message that shows up when cpython cannot be found in a provided path

File: scripts/clib/clib_test_all.sh (hunk 2 above)
Comment (youknowone) [reply scripts/clib/clib_test_all.sh:1]: Just python will be ok.

Comment (MegasKomnenos) [reply scripts/clib/clib_test_all.sh]: > Just python will be ok. We have to fix rustpython to run any python script, eventually.

Done!

**Comment by youknowone** [comment]
Thank you! let me test it a bit.

**Review by youknowone — CHANGES_REQUESTED** [review]

**Comment by youknowone** [comment]
Any suggestion about the script name? `clib` doesn't give good sense what the script does.
We may be able to turn this script to auto-updating trivial cpython library easily.

cc @fanninpm @DimitrisJim

File: scripts/clib/clib_test.py [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+import argparse
+import shutil
+import subprocess
+
+def main():
+    parser = argparse.ArgumentParser(description="Test cpython library")
+    parser.add_argument("-cp", nargs=1, default="../../../cpython", required=False, help="Local cpython path.")
+    parser.add_argument("-rp", nargs=1, default="../..", required=False, help="Local RustPython path.")
+    args = vars(parser.parse_args())
+    CPYTHONPATH = args["cp"]
```
Comment (youknowone) [inline scripts/clib/clib_test.py:11]:
```suggestion
    CPYTHON_PATH = args["cp"]
```

File: scripts/clib/clib_test.py [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+    RPYTHONPATH = args["rp"]
```
Comment (youknowone) [inline scripts/clib/clib_test.py:12]:
```suggestion
    RPYTHON_PATH = args["rp"]
```

File: scripts/clib/clib_test.py [hunk 7]
```diff
@@ -0,0 +1,81 @@
+import os
+import argparse
+import shutil
+import subprocess
+
+def main():
+    parser = argparse.ArgumentParser(description="Test cpython library")
+    parser.add_argument("-cp", nargs=1, default="../../../cpython", required=False, help="Local cpython path.")
```
Comment (youknowone) [inline scripts/clib/clib_test.py:8]:
```suggestion
    parser.add_argument("--cpython", nargs=1, default="../../../cpython", required=False, help="Local cpython path.")
```

unix commonly uses `--` for long argument names.

File: scripts/clib/clib_test.py [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+    parser.add_argument("-rp", nargs=1, default="../..", required=False, help="Local RustPython path.")
```
Comment (youknowone) [inline scripts/clib/clib_test.py:9]:
```suggestion
    parser.add_argument("--rustpython", nargs=1, default="../..", required=False, help="Local RustPython path.")
```

File: scripts/clib/clib_test.py [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+
+    if isinstance(CPYTHONPATH, list):
+        CPYTHONPATH = CPYTHONPATH[0]
+    if isinstance(RPYTHONPATH, list):
+        RPYTHONPATH = RPYTHONPATH[0]
+
+    liblst = list()
+
+    with open("clib_list.txt", 'r') as f:
+        for line in f.readlines():
```
Comment (youknowone) [inline scripts/clib/clib_test.py:22]:
```suggestion
        for line in f:
```
This is just same

File: scripts/clib/clib_test.py [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+            line = line[:line.find('#')]
```
Comment (youknowone) [inline scripts/clib/clib_test.py:23]:
```suggestion
            line = line.split('#')[0]
```

File: scripts/clib/clib_test.py [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+            line = line.strip()
+            if line:
+                liblst.append(line)
+
+    open("clib_out.txt", 'w').close()
```
Comment (youknowone) [inline scripts/clib/clib_test.py:28]: instead of opening outfile, printing to stdout will be ok.
we can pipe it to write it to file

File: scripts/clib/clib_test.py [hunk 12]
```diff
@@ -0,0 +1,81 @@
...
+
+    for lib in liblst:
+        lib_exist = False
+        test_exist = False
+        test_do = True
+        message = [f"{lib}:"]
+
+        if os.path.isfile(f"{RPYTHONPATH}/Lib/{lib}.py"):
+            lib_exist = True
+            os.rename(f"{RPYTHONPATH}/Lib/{lib}.py", f"{RPYTHONPATH}/Lib/{lib}_tmp_mv.py")
```
Comment (youknowone) [inline scripts/clib/clib_test.py:38]: Rather than renaming original library files, redefining `RUSTPYTHONPATH` will be fine. 

This is the logic I think.

1. cp -r Lib LibTest  (but in python, not shell)
1. cp ../cpython/Lib/x LibTest/x
1. RUSTPYTHONPATH=LibTest rustpython Lib/test/test_...

File: scripts/clib/clib_test.py [hunk 13]
```diff
@@ -0,0 +1,81 @@
...
+            message.append(f"No cpython/Lib/test/test_{lib}.py")
+        
+        if test_do:
+            result = subprocess.run(["cargo", "run", "-q", f"{RPYTHONPATH}/Lib/test/test_{lib}.py"], stdout=subprocess.PIPE, stderr=subprocess.STDOUT)
+            result = result.stdout.decode("utf-8")
+
+            if "OK" in result:
+                message.append("OK")
+            else:
+                message.append("Failed")
```
Comment (youknowone) [inline scripts/clib/clib_test.py:62]: maybe you want to define a dataclass and format the result later.

```python
from dataclasses import dataclass

@dataclass
class Item:
    lib_exist: bool
    test_exist: bool
    ...

   def to_string(self):
        return ...
```

File: scripts/clib/clib_test.py (hunk 9 above)
Comment (MegasKomnenos) [reply scripts/clib/clib_test.py:22]: Huh. I did not know that. Thanks!

File: scripts/clib/clib_test.py (hunk 11 above)
Comment (MegasKomnenos) [reply scripts/clib/clib_test.py:28]: Ah, that's a fair point. Yeah, that seems to be a more versatile design.

File: scripts/clib/clib_test.py (hunk 12 above)
Comment (MegasKomnenos) [reply scripts/clib/clib_test.py:38]: Sure, that would be easy. But I am not fully grasping why that would be a better design. Is it because messing with library files via script may open room for possible human mistakes?

File: scripts/clib/readme.md [hunk 14]
```diff
@@ -0,0 +1,9 @@
+To use, open cmd and run clib_test.py with args through python
+and check the result in clib_out.txt
+
+The script assumes that the script is being run from RustPython/scripts/clib,
+and that both RustPython and cpython project directories are located under a same parent directory, aka that they are siblings
+
+If either of those assumptions are false, then you must provide a correct path when running clib_test.py
+
+The script will try to test every component in clib_list.txt
```
[the remark this answers is not in the document]
Comment (MegasKomnenos) [reply scripts/clib/readme.md:9]: I copy-pasted the content of the issue page into a txt file, then used regex to filter out its contents

File: scripts/clib/clib_test.py [hunk 15]
```diff
@@ -0,0 +1,81 @@
+import os
+import argparse
+import shutil
+import subprocess
+
```
[the remark this answers is not in the document]
Comment (MegasKomnenos) [reply scripts/clib/clib_test.py:5]: Sure, that would indeed be a necessary improvement.

File: scripts/clib/clib_test.py [hunk 16]
```diff
@@ -0,0 +1,81 @@
...
+    args = vars(parser.parse_args())
+    CPYTHONPATH = args["cp"]
+    RPYTHONPATH = args["rp"]
+
+    if isinstance(CPYTHONPATH, list):
+        CPYTHONPATH = CPYTHONPATH[0]
+    if isinstance(RPYTHONPATH, list):
+        RPYTHONPATH = RPYTHONPATH[0]
+
+    liblst = list()
```
[the remark this answers is not in the document]
Comment (MegasKomnenos) [reply scripts/clib/clib_test.py:19]: Ah, thanks for catching it. It's a bad habit of mine that I picked up from CP, where we try to shorten variable name as much as possible to the detriment of readability. I try to be self-conscious about it, but it often gets the better of me still.

File: scripts/clib/clib_test.py [hunk 17]
```diff
@@ -0,0 +1,81 @@
...
+            message.append(f"No cpython/Lib/{lib}.py")
+
+        if os.path.isfile(f"{CPYTHONPATH}/Lib/test/test_{lib}.py"):
+            shutil.copyfile(f"{CPYTHONPATH}/Lib/test/test_{lib}.py", f"{RPYTHONPATH}/Lib/test/test_{lib}.py")
+        else:
+            test_do = False
+            message.append(f"No cpython/Lib/test/test_{lib}.py")
+        
+        if test_do:
+            result = subprocess.run(["cargo", "run", "-q", f"{RPYTHONPATH}/Lib/test/test_{lib}.py"], stdout=subprocess.PIPE, stderr=subprocess.STDOUT)
```
[the remark this answers is not in the document]
Comment (MegasKomnenos) [reply scripts/clib/clib_test.py:56]: But wouldn't it be necessary to do cargo build each time as we're changing its library files? Or maybe the interpreter functions independent of its library python files and simply references whatever is written there when it needs to? I'm not familiar with the structure of the project and I assumed the former to be true, but maybe it's not.

File: scripts/clib/clib_test.py (hunk 12 above)
Comment (MegasKomnenos) [reply scripts/clib/clib_test.py:38]: I've tried to use RUSTPYTHONPATH environment variable to control which library files are referenced by rustpython interpreter but I seem to be doing something wrong, as it's working as if it's still referencing files in Lib, not LibTest. Could you guide me through how I could do what you suggested in python?

File: scripts/clib/clib_test.py (hunk 10 above)
Comment (MegasKomnenos) [reply scripts/clib/clib_test.py]: That's a better way to implement comment. Thanks for letting me know!

File: scripts/clib/clib_test.py (hunk 12 above)
Comment (MegasKomnenos) [reply scripts/clib/clib_test.py]: I tried to set environment variable through os.environ. Let me try again with a new method and see how it turns out.

File: scripts/clib/clib_test.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (MegasKomnenos) [reply scripts/clib/clib_test.py]: Huh. I'm seeing an odd change in behavior.  I'm comparing the previous result stored in clib_out.txt with the new result that I'm getting  from this change, and it seems to be slightly different? Some tests that failed before now  are now passing. Odd.

[the remark this answers is not in the document]
Comment (MegasKomnenos) [reply scripts/clib/clib_test.py]: [clib_out_new.txt](https://github.com/RustPython/RustPython/files/11348301/clib_out_new.txt)
[clib_out.txt](https://github.com/RustPython/RustPython/files/11348304/clib_out.txt)

[the remark this answers is not in the document]
Comment (MegasKomnenos) [reply scripts/clib/clib_test.py]: I believe that this command is not working as intended.

result = subprocess.run(
                [RUSTEXECPATH, "-q", self.path_tpython_test],
                stdout=subprocess.PIPE, 
                stderr=subprocess.STDOUT,
                env={"RUSTPYTHONPATH": "/home/kjh/20231/rust/RustPython/LibTest"}
            )

When I run it with the Lib directory removed, it returns ModuleNotFoundError

**Comment by DimitrisJim** [comment]
do we have any updates on the state of this PR? It's been sorta stale for 3 weeks now and we should try and get your work through soon if possible.

**Comment by youknowone** [comment]
My bad, I forgot about this after getting in the parser rally.
The whole idea is awesome. I need to clean up its interface more fit in our other scripts and naming so on.

**Comment by youknowone** [comment]
@DimitrisJim do you have a good name suggestion instead of clib? it is not a C something.
